### PR TITLE
[JENKINS-48761] - Restore binary compatibility with agents running old Remoting versions

### PR DESCRIPTION
**Pull request by oleg-nenashev**
See [JENKINS-48761](https://issues.jenkins-ci.org/browse/JENKINS-48761).

This change reverts new API usages in #3145 and some other places. I made a mistake with this APIs, and my action item is to create a unit test which verifies that the master is at least able to establish connection with an agent with old Remoting version.

### Proposed changelog entries

* Major bug: Restore binary compatibility of Jenkins core logic with old Remoting versions to allow connecting old agents (regression in 2.98)
* ...

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees @jglick

### CODE REVIEW

File: core/src/main/java/hudson/Launcher.java
Comment (jglick) [inline core/src/main/java/hudson/Launcher.java]: typo

File: core/src/main/java/jenkins/security/MasterToSlaveCallable.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/security/MasterToSlaveCallable.java]: This one is 3.11, will need an older API as well

File: core/src/main/java/jenkins/security/MasterToSlaveCallable.java
Comment (jglick) [inline core/src/main/java/jenkins/security/MasterToSlaveCallable.java]: I think you could do something like

```java
@Override
public Channel getChannelOrFail() throws ChannelStateException {
    // copy implementation from Callable
}
```

which would avoid polluting the API with even a restricted function, while still fixing the problem.

**Comment by oleg-nenashev** [comment]
The current code will run on Remoting agents 3.0+ at least, likely on even older ones

**Review by jglick — DISMISSED** [review]
OK as a hotfix.

The broader issue is that we have no static check for the case that a core or plugin remote `Callable` is using an API member introduced in a newer version of Remoting than the agent uses. Not sure what to do about that.

**Comment by oleg-nenashev** [comment]
I could write a FindBugs detector for that, but it's hard do guarantee full
coverage. Maybe we should start from playing with createOnlineSlave() for
old Remoting versions and duplicate some tests

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers does it deserve an out-of-order release? I'd guess so, but maybe it makes sense to take another day for test automation, etc.

**Comment by jglick** [comment]
> playing with createOnlineSlave() for old Remoting versions

Would be tricky since you `remoting.jar` is picked up from the Maven test classpath, which cannot contain multiple versions. Better would be to use `docker-fixtures` from `jenkinsci/jenkins/test/`, which would allow us to connect to an actual agent running in another environment and using a pinned version of Remoting. (For that matter, did none of the existing tests using Dockerized SSH agents in `acceptance-test-harness` catch this problem?)

> does it deserve an out-of-order release?

I am not sure. It only affects people running old agent versions. If they are going to bother updating core to get an out-of-order release, maybe they could just as easily update their agent. No strong opinion.

**Comment by oleg-nenashev** [comment]
> For that matter, did none of the existing tests using Dockerized SSH agents in acceptance-test-harness catch this problem?

Let's see. But we do not run ATH before weeklies now

**Comment by oleg-nenashev** [comment]
@jglick After some consideration I decided to write a test without Docker containers. It required doing some foundation work for [JENKINS-48766](https://issues.jenkins-ci.org/browse/JENKINS-48766), but I think we need to do it anyway.

File: test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java
Comment (oleg-nenashev) [inline test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java]: This test fails on Jenkins 2.99/98. `ChannelClosedException` API compatibility is not tested though

File: pom.xml
Comment (oleg-nenashev) [inline pom.xml]: Just picked first version with excessive stability improvements and available changelog: https://github.com/jenkinsci/remoting/blob/master/CHANGELOG.md#260 . This version has been released on June 10, 2016 => old enough.

IMHO the version should be bumped to 3.0 once https://issues.jenkins-ci.org/browse/JENKINS-48766 is implemented. It would unblock #3188 from @Vlatombe

**Comment by oleg-nenashev** [comment]
@reviewbybees Needs re-reviews (mostly tests and Restricted logic in the core for enabling such kind of tests)

File: core/src/main/java/jenkins/security/MasterToSlaveCallable.java
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/security/MasterToSlaveCallable.java]: I was not sure it's 100% safe for defaults and decided not to try due to the urgency. Can try it now since I have a test which should reveal the problem if any

File: core/src/filter/resources/jenkins/slaves/remoting-info.properties
Comment (jglick) [inline core/src/filter/resources/jenkins/slaves/remoting-info.properties]: Or this could be computed from `remoting*.jar!/META-INF/MANIFEST.MF#Version` using `Which`.

File: core/src/filter/resources/jenkins/slaves/remoting-info.properties
Comment (jglick) [inline core/src/filter/resources/jenkins/slaves/remoting-info.properties]: s/minimal/minimum/g

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: typo

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: Restricted, so delete.

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: No need to check for null. If the file is missing, core is corrupt, so die now.

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: Ditto—just rethrow as `ExceptionInInitializerError`.

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: Throw an `IllegalStateException`.

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: ditto

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: Ditto. We are in full control of this file, so there is no reason to tolerate any errors.

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [inline core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: `@Nonnull` since we know these will be defined.

File: core/src/test/java/jenkins/slaves/RemotingVersionInfoTest.java
Comment (jglick) [inline core/src/test/java/jenkins/slaves/RemotingVersionInfoTest.java]: Unnecessary if they are just `@Nonnull` to begin with, since other tests would die if someone screwed this up.

File: test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java
Comment (jglick) [inline test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java]: BTW as a stylistic matter it is safest to use `OldRemotingAgentTest.class` in place of `getClass()`. (Unless the test suite is `final` in which case they are equivalent.)

File: test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java
Comment (jglick) [inline test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java]: This is test code. Just throw it up!

Or, better, make that method `@Restricted(NoExternalUse.class) public`.

File: core/src/filter/resources/jenkins/slaves/remoting-info.properties
Comment (oleg-nenashev) [reply core/src/filter/resources/jenkins/slaves/remoting-info.properties]: I would prefer to keep it as is. 
 
* At some point we may want to write a "Remoting Compatibility Tester" tool or so
* Not sure how which is going to behave once we have pluggable core components in place

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: There is a TODO for making it public above

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: The implementation is similar to `jenkins.model.Jenkins`'s appoarch to reading `jenkins-version.properties`. I agree this should not be a problem, but there may be edge cases (e.g. custom packaging). I see no problem with having a fault-tolerant code since it is a regression fix

File: test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java
Comment (oleg-nenashev) [reply test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java]: yup

File: test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java
Comment (oleg-nenashev) [reply test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java]: Right, it should never happen

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [reply core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: In the case of custom packaging, it is the responsibility of the packager to not make mistakes. If they do, their WAR will be broken. We should not engineer in robustness for something that is not a user error and is trivially detected.

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (jglick) [reply core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: So if and when that happens, you will add a `@since` with _that_ version, as with adding any new API. We should only have `@since TODO` comments in the code base for those things which we have forgotten to fill in after merging.

File: core/src/main/java/jenkins/security/MasterToSlaveCallable.java
Comment (jglick) [reply core/src/main/java/jenkins/security/MasterToSlaveCallable.java]: Not at all urgent, just an idea which might simplify the patch.

File: core/src/main/java/jenkins/slaves/RemotingVersionInfo.java
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: Hmm, another potential use-case is calling this class from the agent. I doubt we would be dare enough to send Jenkins core over Remoting. Do we do that in the case of such requests? I hope not

Comment (oleg-nenashev) [reply core/src/main/java/jenkins/slaves/RemotingVersionInfo.java]: Let's keep it as is. I will improve checks when exposing it to public API

**Comment by oleg-nenashev** [comment]
@jglick addressed some comments. One biggest open question is `RemotingVersionInfo`. You propose to make the checks stricter and to throw exceptions on initialization errors. I kinda agree with that, but I propose to postpone it and to handle it in https://issues.jenkins-ci.org/browse/JENKINS-48766 while making the API public. 

Now it's used only for tests, and I would prefer to keep it permissive in order to avoid new regressions.

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java]: unrelated but OK

File: test/pom.xml
Comment (jglick) [inline test/pom.xml]: BTW may want to s/minimal/minimum/ here too for consistency.

**Review by jglick — APPROVED** [review]

File: test/pom.xml
Comment (oleg-nenashev) [reply test/pom.xml]: Will address it in a follow-up

**Comment by oleg-nenashev** [comment]
@reviewbybees done
@jenkinsci/code-reviewers I kindly ask for approval to got forward with merge and out-of-order release. Community stats are bad due to that regression:

screen shot 2018-01-03 at 19 24 19

**Review by abayer — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Merging it, this change will unlikely make the situation worse

**Comment by oleg-nenashev** [comment]
Retrospective thread: https://groups.google.com/forum/#!topic/jenkinsci-dev/-VQ5YU0yYhg

**Pull request merged**
